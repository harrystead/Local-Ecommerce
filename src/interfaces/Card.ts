export interface ICard {
  ProductNo: string;
  Title: string;
  Description: string;
  Price: IPrice;
  ProductImage: IProductImage;
  Reviews: IReviews;
}

interface IProductImageLink {
  Href: string;
  Method: string;
  Rel: string;
  Title: string;
}

interface IProductImage {
  Link: IProductImageLink;
  MimeType: string;
}

interface IPrice {
  Value: number;
  Currency: string;
}

interface IReviews {
  AverageStarReviewRating: number;
  ReviewCount: number;
}
