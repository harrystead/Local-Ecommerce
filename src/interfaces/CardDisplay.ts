export interface ICardDisplay {
  ThumbnailUrl: string;
  Title: string;
  Price: number;
}
