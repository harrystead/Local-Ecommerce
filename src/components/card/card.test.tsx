import { render, fireEvent } from '@testing-library/react';
import { ICard } from 'interfaces/Card';
import { useNavigate } from 'react-router-dom';
import Card from './card';

jest.mock('react-router-dom', () => ({
  useNavigate: jest.fn(),
}));

describe('card component', () => {
  const generateMockCard = (overrides: Partial<ICard> = {}): ICard => ({
    ProductNo: '12345',
    ProductImage: {
      Link: {
        Href: 'https://example.com/image.jpg',
        Title: 'Image Title',
        Rel: 'hello',
        Method: 'GET',
      },
      MimeType: 'hello',
    },
    Title: 'Card Title',
    Price: {
      Value: 4.99,
      Currency: 'Â£',
    },
    Description: 'Card Description',
    Reviews: {
      AverageStarReviewRating: 4.0,
      ReviewCount: 3,
    },
    ...overrides,
  });

  const mockNavigate = jest.fn();

  beforeEach(() => {
    (useNavigate as jest.Mock).mockReturnValue(mockNavigate);
    mockNavigate.mockClear();
  });

  it('renders correctly', () => {
    const mockCard = generateMockCard();
    const { getByRole } = render(<Card card={mockCard} />);
    const cardImage = getByRole('img');

    expect(cardImage).toHaveAttribute('src', mockCard.ProductImage.Link.Href);
    expect(cardImage).toHaveAttribute('alt', mockCard.ProductImage.Link.Title);
  });

  it('calls navigate with correct argument on click', () => {
    const mockCard = generateMockCard();
    const { getByRole } = render(<Card card={mockCard} />);
    const cardImage = getByRole('img');

    fireEvent.click(cardImage);
    expect(mockNavigate).toHaveBeenCalledWith(mockCard.ProductNo);
  });

  it('matches snapshot', () => {
    const mockCard = generateMockCard();
    const { asFragment } = render(<Card card={mockCard} />);
    expect(asFragment()).toMatchSnapshot();
  });

  it('renders correctly with different props', () => {
    const mockCard = generateMockCard({
      ProductNo: '123452',
      ProductImage: {
        Link: {
          Href: 'https://example2.com/image.jpg',
          Title: 'Image Title2',
          Method: 'get',
          Rel: 'hey',
        },
        MimeType: 'dedmed',
      },
    });

    const { getByRole } = render(<Card card={mockCard} />);
    const cardImage = getByRole('img');

    expect(cardImage).toHaveAttribute('src', mockCard.ProductImage.Link.Href);
    expect(cardImage).toHaveAttribute('alt', mockCard.ProductImage.Link.Title);
  });

  it('does not navigate if productNo is not provided', () => {
    const faultyMockCard = generateMockCard({
      ProductNo: '',
    });

    const { getByRole } = render(<Card card={faultyMockCard} />);
    const cardImage = getByRole('img');

    fireEvent.click(cardImage);
    expect(mockNavigate).not.toHaveBeenCalled();
  });

  it('renders without image data', () => {
    const faultyMockCard = generateMockCard({
      ProductImage: {
        Link: {
          Href: '',
          Title: '',
          Method: '',
          Rel: '',
        },
        MimeType: '',
      },
    });

    const { queryByRole } = render(<Card card={faultyMockCard} />);
    const cardImage = queryByRole('img');

    expect(cardImage).toBeNull();
  });
});
