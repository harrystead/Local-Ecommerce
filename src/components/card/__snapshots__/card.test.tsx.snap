// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Card component matches snapshot 1`] = `
<DocumentFragment>
  <div>
    <img
      alt="Image Title"
      class="card"
      src="https://example.com/image.jpg"
    />
  </div>
</DocumentFragment>
`;
